app/components/NewChannelItem/index.jsx

```diff
@@ -24,19 +28,28 @@ export default class ChannelItem extends React.Component {
     newChannelInput.focus();
   }
 
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return !(
+      Immutable.is(nextProps.channels, this.props.channels) &&
+      nextState.validationMessage === this.state.validationMessage &&
+      nextState.infoType === this.state.infoType
+    );
+  }
+
+
   onKeyPress = (e) => {
-    if (e.charCode !== 13) {
-      return;
+    if (e.which === 13) {
+      this.addChannel();
     }
-    this.addChannel();
   }
 
 
   addChannel = () => {
     const channelName = this.state.channelName.trim();
-    const sameChannel = this.props.channels.find(channel => channel.get('name') === channelName);
-    if (channelName && !sameChannel) {
-      this.props.replaceDirtyChannel(new Map({name: this.state.channelName, isDirty: true}));
+    const hasError = this.validateChannelName(channelName);
+    if (!hasError) {
+      this.props.replaceDirtyChannel(new Map({name: channelName, isDirty: true}));
       this.props.newChannel({name: channelName});
       const newChannelItem = this.refs.newChannel.getDOMNode();
       newChannelItem.className = newChannelItem.className + ' new-channel_disabled';
```

Comment:
Where did you see the angular in this code? I run a validate function and check results. If there are no errors i run a addChannel action.